src/main/scala/org/graphframes/lib/AggregateMessages.scala

```diff
@@ -100,22 +101,50 @@ class AggregateMessages private[graphframes] (private val g: GraphFrame)
    *   aggResult.join(g.vertices, ID)
    * }}}
    */
-  def agg(aggCol: Column): DataFrame = {
+  def agg(aggCol: Column, aggCols: Column*): DataFrame = {
+    def containsStructNameForSpark2(columnName: String, structName: String) =
+      columnName.startsWith(s"${structName}.")
+    def containsStructNameForSpark16(columnName: String, structName: String) =
+      columnName.startsWith(s"${structName}[")
+    def removeStructNameForSpark2(columnName: String, structName: String) =
+      columnName.substring(s"${structName}.".length)
+    def removeStructNameForSpark16(columnName: String, structName: String) =
+      columnName.substring(s"${structName}[".length, columnName.length - 1)
+
+    def removeStructName(columnName: String) = columnName match {
+      case cn if containsStructNameForSpark16(cn, s"${GraphFrame.SRC}") =>
+        removeStructNameForSpark16(cn, s"${GraphFrame.SRC}")
+      case cn if containsStructNameForSpark16(cn, s"${GraphFrame.DST}") =>
+        removeStructNameForSpark16(cn, s"${GraphFrame.DST}")
+      case cn if containsStructNameForSpark16(cn, s"${GraphFrame.EDGE}") =>
+        removeStructNameForSpark16(cn, s"${GraphFrame.EDGE}")
+      case cn if containsStructNameForSpark2(cn, s"${GraphFrame.SRC}") =>
+        removeStructNameForSpark2(cn, s"${GraphFrame.SRC}")
+      case cn if containsStructNameForSpark2(cn, s"${GraphFrame.DST}") =>
+        removeStructNameForSpark2(cn, s"${GraphFrame.DST}")
+      case cn if containsStructNameForSpark2(cn, s"${GraphFrame.EDGE}") =>
+        removeStructNameForSpark2(cn, s"${GraphFrame.EDGE}")
+      case cn  => cn
+    }
+    def msgColumn(df: DataFrame) = df.columns.filter(_ != ID) match {
+      case Array(c) => df.withColumnRenamed(c, AggregateMessages.MSG_COL_NAME)
+      case columns => df.select(
+        df(ID),
+        struct(columns.sorted.map(c => col(s"`${c}`").as(removeStructName(c))) :_*)
+          .as(AggregateMessages.MSG_COL_NAME))
```

Comment:
this is the way AggregationMessages works: starts with a DataFrame resulting from calling GraphFrame.triplets, which generates a DF with columns: "src", "dst", "edge", each of which is of type struct with one struct element for each original attribute.

When 1 attribute per AggregateMessages was supported, this attribute was used in the projection of a query to this df. So let's say I have the following graph:

```GraphFrame(v:[id: string, name: string, age: int], e:[src: string, dst: string, relationship: string])```

If I wanted to send vertex attribute "name" from src to dst, that would generate the following query:

```select dst['id'] AS id, src['name'] AS MSG from df.triplets```

in the inverse direction

```select src['id'] AS id, dst['name'] AS MSG from df.triplets```

and in both directions:

```select dst['id'] AS id, src['name'] AS MSG from df.triplets UNION select src['id'] AS id, dst['name'] AS MSG from df.triplets```

Now, with the changes I introduced, multiples attribute can be sent in a message. The way it is resolved is to group all the attributes in a struct type. So following the same example, if I wanted to send both attributes "name" and "age" from src to dst the new query would be:

```select dst['id'] AS id, struct(src['name'], src['age']) AS MSG from df.triplets```

and similar for the other directions.

So this solution brings some challenges:

When I try to create a struct from existing struct elements the new generated df creates a struct naming each element col1, col2.. colN rather that the original name. With the example above try the following:

```
g.triplets.select(struct(expr("src['id']")).as("MSG")).printSchema
g.triplets.select(struct(col("src")("id")).as("MSG")).printSchema
```
and you will see what I'm saying. So what basically line 133 does is to create the struct but maintaining the original attribute names.

Additionally what removeStructName and the other helper functions do is the following: If I kept the original name as I said above, then I would end up with the following struct:

```struct("src.name", "src.age") as MSG```

that represents the message sent from src to dst. However, the message in the opposite direction would be:

```struct("dst.name", "dst.age") as MSG```

Since both structs have different element names, the union between both df doesn't work. So the helper method removes the struct name from the columns to end up having the following struct for both directions:

```struct("name", "age") as MSG```

This way the union is possible.

Last but not least, what I found is that while in Spark 1.6 the default column name set in a DF when you select a struct element is ```struct_name[struct_element]```, i.e. ```src[name]``` in Spark 2.x it is ```struct_name.struct_element```, i.e. ```src.name```. That's why you will see the functions removeStructNameForSpark2 and removeStructNameForSpark16

Hope this clarifies your doubts. Please let me know if you have further questions